### ISSUE TITLE
Selenide creates a screenshot however does not report it in exception log

### ISSUE THREAD

**Issue by dmak** [selenide/selenide#234]
The Unit test generates the following trace:

```
Running org.integration.MyTest
10.21 21:01:40.267 ...
...
10.21 21:06:39.267 ...
test(org.integration.MyTest)  Time elapsed: 48.189 sec  <<< FAILURE!
com.codeborne.selenide.ex.ElementNotFound: Element not found {By.xpath: //table/tbody/tr/a='']}
Expected: not hidden
    at sun.reflect.GeneratedConstructorAccessor12.newInstance(Unknown Source)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
    at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
    at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
    at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
    at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:348)
    at org.openqa.selenium.remote.RemoteWebDriver.findElementByXPath(RemoteWebDriver.java:445)
    at org.openqa.selenium.By$ByXPath.findElement(By.java:358)
    at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:340)
    at com.codeborne.selenide.impl.WebElementSelector.findElement(WebElementSelector.java:28)
    at com.codeborne.selenide.impl.WaitingSelenideElement.getActualDelegate(WaitingSelenideElement.java:56)
    at com.codeborne.selenide.impl.AbstractSelenideElement.checkCondition(AbstractSelenideElement.java:528)
    at com.codeborne.selenide.impl.AbstractSelenideElement.should(AbstractSelenideElement.java:517)
    at com.codeborne.selenide.impl.AbstractSelenideElement.invokeShould(AbstractSelenideElement.java:305)
    at com.codeborne.selenide.impl.AbstractSelenideElement.dispatchSelenideMethod(AbstractSelenideElement.java:179)
    at com.codeborne.selenide.impl.AbstractSelenideElement.dispatchAndRetry(AbstractSelenideElement.java:90)
    at com.codeborne.selenide.impl.AbstractSelenideElement.invoke(AbstractSelenideElement.java:65)
    at com.codeborne.selenide.impl.WaitingSelenideElement.invoke(WaitingSelenideElement.java:19)
    at com.sun.proxy.$Proxy7.waitUntil(Unknown Source)
```

The test also generates two files:

```
$ ls -l
 22072 Oct 21 21:03 1445454210701.0.html
 97443 Oct 21 21:03 1445454210701.0.png
```

As stack trace is printed by JUnit after all tests are executed in one given test case (= class), I know only that test case was running between 21:01:40 and 21:06:39. File creation times fit that period, So given files should have been generated by this test case. However their names are not present in the stack trace.

Expected: If screenshot is taken, it's name should appear in error dump. Otherwise screenshot should not be taken (it was created by mistake).

Note: Only one test (listed above) has failed in whole test suit. And there are only two given files generated for given date (21 Oct), so it couldn't be that they belong to another test run.

**Comment by asolntsev** [comment]
@dmak  hm... magic... I need to investigate it.

**Comment by dmak** [comment]
I think that is somehow connected to `UIAssertionError#collectCurrentWebdriverData()` which is called only from `wrap()`. So perhaps the screenshot is created, but not registered in `UIAssertionError`.

**Comment by asolntsev** [comment]
@dmak Please provide exact test code.
Are you using method `shouldNoBe(hidden)`? Are you searching element by xpath `//table/tbody/tr/a=''` like this?

```java
  @Test
  public void testDMak() {
    $(By.xpath("//table/tbody/tr/a='']")).shouldNotBe(hidden);
  }
```
Actually this is **invalid XPath**, and I am getting the following error for this code:

```java
org.openqa.selenium.InvalidSelectorException: invalid selector: Unable to locate an element with the xpath expression //table/tbody/tr/a=''] because of the following error:
SyntaxError: Failed to execute 'evaluate' on 'Document': The string '//table/tbody/tr/a='']' is not a valid XPath expression.
  (Session info: chrome=47.0.2526.106)
  (Driver info: chromedriver=2.20.353124 (035346203162d32c80f1dce587c8154a1efa0c3b),platform=Mac OS X 10.11.2 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 33 milliseconds
For documentation on this error, please visit: http://seleniumhq.org/exceptions/invalid_selector_exception.html
Build info: version: '2.48.2', revision: '41bccdd10cf2c0560f637404c2d96164b67d9d67', time: '2015-10-09 13:08:06'
System info: host: 'Andreis-MacBook-Pro-2.local', ip: '127.0.0.1', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.2', java.version: '1.7.0_45'
*** Element info: {Using=xpath, value=//table/tbody/tr/a='']}
Session ID: 257a8460cb9ec7d06fc0a92d7e593734
Driver info: org.openqa.selenium.chrome.ChromeDriver
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:647)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:353)
	at org.openqa.selenium.remote.RemoteWebDriver.findElementByXPath(RemoteWebDriver.java:490)
	at org.openqa.selenium.By$ByXPath.findElement(By.java:361)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:345)
	at com.codeborne.selenide.impl.WebElementSelector.findElement(WebElementSelector.java:28)
	at com.codeborne.selenide.impl.ElementFinder.getWebElement(ElementFinder.java:55)
	at com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:44)
	at com.codeborne.selenide.commands.ShouldNot.shouldNot(ShouldNot.java:36)
	at com.codeborne.selenide.commands.ShouldNot.execute(ShouldNot.java:29)
	at com.codeborne.selenide.commands.ShouldNot.execute(ShouldNot.java:12)
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:135)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchSelenideMethod(SelenideElementProxy.java:119)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:80)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:56)
	at com.sun.proxy.$Proxy10.shouldNotBe(Unknown Source)
	at integration.DMakTest.test1(DMakTest.java:32)

**Comment by dmak** [comment]
Sorry, I have cropped xPath and now it is incorrect.
`//table/tbody/tr[td/a='some_text']`
(select a table row with link "some_text").

**Comment by dmak** [comment]
I think there is something wrong with the exception handler. Example with `org.openqa.selenium.NoSuchElementException` which does not generate screenshot:

```
test_176(org.mycompany.Test)  Time elapsed: 60.241 sec  <<< FAILURE!
com.codeborne.selenide.ex.ElementNotFound: Element not found {#tab-area li[aria-controls='08016978']}
Expected: css class 'done'
	at sun.reflect.GeneratedConstructorAccessor9.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:348)
	at org.openqa.selenium.remote.RemoteWebDriver.findElementByCssSelector(RemoteWebDriver.java:437)
	at org.openqa.selenium.By$ByCssSelector.findElement(By.java:426)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:340)
	at com.codeborne.selenide.impl.WebElementSelector.findElement(WebElementSelector.java:28)
	at com.codeborne.selenide.impl.WaitingSelenideElement.getActualDelegate(WaitingSelenideElement.java:56)
	at com.codeborne.selenide.impl.AbstractSelenideElement.checkCondition(AbstractSelenideElement.java:528)
	at com.codeborne.selenide.impl.AbstractSelenideElement.should(AbstractSelenideElement.java:517)
	at com.codeborne.selenide.impl.AbstractSelenideElement.invokeShould(AbstractSelenideElement.java:305)
	at com.codeborne.selenide.impl.AbstractSelenideElement.dispatchSelenideMethod(AbstractSelenideElement.java:179)
	at com.codeborne.selenide.impl.AbstractSelenideElement.dispatchAndRetry(AbstractSelenideElement.java:90)
	at com.codeborne.selenide.impl.AbstractSelenideElement.invoke(AbstractSelenideElement.java:65)
	at com.codeborne.selenide.impl.WaitingSelenideElement.invoke(WaitingSelenideElement.java:19)
	at com.sun.proxy.$Proxy7.waitUntil(Unknown Source)
	at org.mycompany.CommonUtils.waitForPanelToLoad(CommonUtils.java:171)
Caused by: org.openqa.selenium.NoSuchElementException: Unable to locate element: {"method":"css selector","selector":"#tab-area li[aria-controls='08016978']"}
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.47.1', revision: 'unknown', time: '2015-07-30 11:02:44'
System info: host: 'luu128d', ip: '10.14.1.85', os.name: 'Linux', os.arch: 'amd64', os.version: '3.0.101-63-default', java.version: '1.7.0_25'
Driver info: driver.version: unknown
	at <anonymous class>.FirefoxDriver.prototype.findElementInternal_(file:///tmp/anonymous8049683132435027632webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:10667)
	at <anonymous class>.FirefoxDriver.prototype.findElement(file:///tmp/anonymous8049683132435027632webdriver-profile/extensions/fxdriver@googlecode.com/components/driver-component.js:10676)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_/h(file:///tmp/anonymous8049683132435027632webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12643)
	at <anonymous class>.DelayedCommand.prototype.executeInternal_(file:///tmp/anonymous8049683132435027632webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12648)
	at <anonymous class>.DelayedCommand.prototype.execute/<(file:///tmp/anonymous8049683132435027632webdriver-profile/extensions/fxdriver@googlecode.com/components/command-processor.js:12590)

```

**Comment by kasparjarek** [comment]
I have same problem, when I use maven to run tests.

I noticed that there are differences when I use or do not use following rule (although the documentation says 'Actually it’s rudiment. You don’t need it, because Selenide does it automatically.').
```java
import com.codeborne.selenide.junit.ScreenShooter;

@Rule
public ScreenShooter makeScreenshotOnFailure = ScreenShooter.failedTests();
```

**Behaviour withou Rule:**
* All screenshots are saved in the same directory
* Screenshots are probably captured just when UIAssertionError occurs. For example:
  * This code capture screenshot
  ```java
$(".foo").shouldBe(visible);
  ```
  * This code does not capture screenshot
  ```java
import org.junit.Assert;
...
Assert.assertTrue(false);
  ```

**Behavior with Rule:**
* Screenshots are saved in separate directories according to package, test class and test name
* Screenshots are captured always when test fails (even when junit assert fails)
  * Test fails caused by junit assert acctualy have screenshot path metioned in stack trace

Using selenide version 3.4

**Comment by TMartysiuk** [comment]
I have the same problem: when some step fails screenshots are created but not printed in the logs. There are no “Timeout” or “Caused by” entries either.  
It looks OK on previous version which we used - Selenide 2.14 
Now we use Selenide v3.4 + Maven + easyb.
Can someone please help me to resolve this problem?

**Comment by asolntsev** [comment]
@TMartysiuk @dmak @kasparjarek 
Sorry for the long delay.
I guess you don't see “Timeout” or “Caused by" in the **maven output**. 
But when you open a test report, you will see “Timeout” or “Caused by” there. Can you double-check this?

**Comment by TMartysiuk** [comment]
Hi, 
I checked this, and in maven output we don't have this information. In our test reports  (easyb html report and surefire xml report) we don't have this information either. But I've created a junit test case and then just run it in Eclipse, and in junit failed trace I can see this missing information. 
I understand, that the problem is propably in maven. But anyway, thanks for response

**Comment by asolntsev** [comment]
@TMartysiuk Can you share some working example with Selenide + EasyB?
Probably create a sample project in github.com/selenide-examples/ ?

**Comment by BorisOsipov** [comment]
closed due no reproducable example.

**Comment by svschouw-bb** [comment]
Not sure if anyone still watches this, and whether this was the case when this bug was reported, but I think the problem is that the screenshot information is only shown in `UIAssertionError.toString()`, but not in `UIAssertionError.getMessage()`. And maven only shows the `.getMessage()` and (partial) stacktrace.

**Comment by asolntsev** [comment]
@svschouw-bb Wow. Thank you for the information. 

Does it mean that the screenshot is not visible **only in maven console output**, but **is still visible** in the html test report generated by maven?

**Comment by svschouw-bb** [comment]
No, it looks like maven only ever uses the `.getMessage()`, never the `.toString()`.

**Comment by asolntsev** [comment]
@svschouw-bb Wow. That was unexpected!
Yes, you are right. Maven doesn't use `Throwable.toString()`, only `Throwable.getMessage()`. 

Well, we need to refactor Selenide exceptions a little bit to include all the required information into `Throwable.getMessage()`. Reopening the issue.

**Comment by asolntsev** [comment]
@svschouw-bb @dmak As a result of my investigation, I found that Maven has changed behaviour meanwhile. 
1. My current Maven version "3.6.0" by default uses `maven-surefire-plugin:2.12.4` which uses `e.toString()` (which contains screenshot)
2. But if you explicitly declare in your pom.xml latest version `maven-surefire-plugin:2.22.2`, then it uses `e.getClass() + e.getMessage()` instead of `e.toString()`. And it doesn't contain a screenshot anymore. :(

I will prepare a pull request, so that Selenide will always include screenshot in `e.getMessage()`, not only `e.toString()`.

**Comment by asolntsev** [comment]
@svschouw-bb @dmak fixed in Selenide 5.3.1
